File: UnionFind/Cpp/src/union_find.cpp
#include "union_find.h"

UnionFind::UnionFind(usize length) noexcept: father(std::make_unique<usize[]>(length)), length(length) {
    for (usize i = 0; i < length; i++) {
        father[i] = i;
    }
}

void UnionFind::merge(usize x, usize y) noexcept {
    if (!is_connected(x, y)) {
        father[find_root(x)] = find_root(y);
    }
}

bool UnionFind::is_connected(usize x, usize y) noexcept {
    return find_root(x) == find_root(y);
}

UnionFind::usize UnionFind::size() const noexcept {
    return length;
}

UnionFind::usize UnionFind::num_blocks() noexcept {
    auto visited = std::make_unique<bool[]>(length);
    usize res = 0;
    for (usize i = 0; i < length; i++) {
        usize root = find_root(i);
        if (!visited[root]) {
            res++;
            visited[root] = true;
        }
    }
    return res;
}

UnionFind::usize UnionFind::find_root(usize x) noexcept {
    if (father[x] == x) {
        return x;
    }
    usize temp = find_root(father[x]);
    father[x] = temp;
    return temp;
}

File: UnionFind/Java/src/UnionFind.java
public class UnionFind {
    private final int[] father;

    public UnionFind(int length) {
        father = new int[length];
        for (int i = 0; i < length; i++) {
            father[i] = i;
        }
    }

    public void merge(int x, int y) {
        if (!isConnected(x, y)) {
            father[findRoot(x)] = findRoot(y);
        }
    }

    public boolean isConnected(int x, int y) {
        return findRoot(x) == findRoot(y);
    }

    public int size() {
        return father.length;
    }

    public int numBlocks() {
        boolean[] visited = new boolean[father.length];
        int res = 0;
        for (int i = 0; i < father.length; i++) {
            int root = findRoot(i);
            if (!visited[root]) {
                res++;
                visited[root] = true;
            }
        }
        return res;
    }

    private int findRoot(int x) {
        if (father[x] == x) {
            return x;
        }
        int temp = findRoot(father[x]);
        father[x] = temp;
        return temp;
    }
}
